text/chapter8.md

```diff
@@ -503,124 +510,163 @@ The `ST` effect is used to manipulate mutable state. As pure functional programm
 
 ```haskell
-simulate' :: Number -> Number -> Number -> Number
-simulate' x0 v0 time = runPure (runST (simulate x0 v0 time))
+simulate' :: Number -> Number -> Int -> Number
+simulate' x0 v0 time = run (simulate x0 v0 time)
 ```
 
 You can even try running this function in PSCi:
 
 ```text
 > import Main
 
-> simulate' 100.0 0.0 0.0
+> simulate' 100.0 0.0 0
 100.00
 
-> simulate' 100.0 0.0 1.0
+> simulate' 100.0 0.0 1
 95.10
 
-> simulate' 100.0 0.0 2.0
+> simulate' 100.0 0.0 2
 80.39
 
-> simulate' 100.0 0.0 3.0
+> simulate' 100.0 0.0 3
 55.87
 
-> simulate' 100.0 0.0 4.0
+> simulate' 100.0 0.0 4
 21.54
 ```
 
-In fact, if we inline the definition of `simulate` at the call to `runST`, as follows:
+In fact, if we inline the definition of `simulate` at the call to `run`, as follows:
 
 ```haskell
 simulate :: Number -> Number -> Int -> Number
-simulate x0 v0 time = runPure $ runST do
-  ref <- newSTRef { x: x0, v: v0 }
-  forE 0 (time * 1000) \_ -> do
-    modifySTRef ref \o ->
-      { v: o.v - 9.81 * 0.001
-      , x: o.x + o.v * 0.001
-      }
-    pure unit
-  final <- readSTRef ref
-  pure final.x
+simulate x0 v0 time =
+  run do
+    ref <- new { x: x0, v: v0 }
+    for 0 (time * 1000) \_ ->
+      modify
+        ( \o ->
+            { v: o.v - 9.81 * 0.001
+            , x: o.x + o.v * 0.001
+            }
+        )
+        ref
+    final <- read ref
+    pure final.x
 ```
 
 then the compiler will notice that the reference cell is not allowed to escape its scope, and can safely turn it into a `var`. Here is the generated JavaScript for the body of the call to `runST`:
 
 ```javascript
-var ref = { x: x0, v: v0 };
-
-Control_Monad_Eff.forE(0)(time * 1000 | 0)(function (i) {
-  return function __do() {
-    ref = (function (o) {
-      return {
-        v: o.v - 9.81 * 1.0e-3,
-        x: o.x + o.v * 1.0e-3
-      };
-    })(ref);
-    return Prelude.unit;
-  };
-})();
-
-return ref.x;
-```
-
-The `ST` effect is a good way to generate short JavaScript when working with locally-scoped mutable state, especially when used together with actions like `forE`, `foreachE`, `whileE` and `untilE` which generate efficient loops in the `Eff` monad.
+var simulate = function (x0) {
+    return function (v0) {
+        return function (time) {
+            return (function __do() {
+                var ref = {
+                    value: {
+                        x: x0,
+                        v: v0
+                    }
+                };
+                Control_Monad_ST_Internal["for"](0)(time * 1000 | 0)(function (v) {
+                    return Control_Monad_ST_Internal.modify(function (o) {
+                        return {
+                            v: o.v - 9.81 * 1.0e-3,
+                            x: o.x + o.v * 1.0e-3
+                        };
+                    })(ref);
+                })();
+                return ref.value.x;
+            })();
+        };
+    };
+};
```

Comment:
This is not as concise as is was in the original.

Reply:
Should the new version have a call to `Control_Monad_ST_Internal.modify`?

This is the [original](https://github.com/paf31/purescript-book/blob/master/text/chapter8.md#mutable-state) body:
```js
var ref = { x: x0, v: v0 };     
                                  
Control_Monad_Eff.forE(0)(time * 1000 | 0)(function (i) {    
  return function __do() {    
    ref = (function (o) {    
      return {                                                                         
        v: o.v - 9.81 * 1.0e-3,                                               
        x: o.x + o.v * 1.0e-3    
      };    
    })(ref);    
    return Prelude.unit;    
  };    
})();    
                                       
return ref.x;
```
This is the new body:
```js
var ref = { value: { x: x0, v: v0 } };                
 
Control_Monad_ST_Internal["for"](0)(time * 1000 | 0)(function (v) {                
  return Control_Monad_ST_Internal.modify(function (o) {                  
    return {                    
      v: o.v - 9.81 * 1.0e-3,                      
      x: o.x + o.v * 1.0e-3                      
    };                    
  })(ref);                  
})();
```

Reply:
Yes, this probably does indicate a problem. There is an inliner rule for `modify` built in to the compiler but it looks like it is no longer firing for some reason.